src/com/sun/jna/Cleaner.java

```diff
@@ -0,0 +1,152 @@
+ *
+ * You can freely decide which license you want to apply to
+ * the project.
+ *
+ * You may obtain a copy of the LGPL License at:
+ *
+ * http://www.gnu.org/licenses/licenses.html
+ *
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "LGPL2.1".
+ *
+ * You may obtain a copy of the Apache License at:
+ *
+ * http://www.apache.org/licenses/
+ *
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "AL2.0".
+ */
+
+package com.sun.jna;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Implement ReferenceQueue based cleanup of resources associated with GCed
+ * objects. It replaces the {@code Object#finalize} based resource deallocation
+ * that is deprecated for removal from the JDK.
+ *
+ * <p><strong>This is not intended for public use.</strong> On JDK 9+ use
+ * {@link java.lang.ref.Cleaner}.</p>
+ */
+public class Cleaner {
+    private static final Cleaner INSTANCE = new Cleaner();
+
+    static Cleaner getCleaner() {
+        return INSTANCE;
+    }
+
+    private final ReferenceQueue<Object> referenceQueue;
+    private final Thread cleanerThread;
+    private CleanerRef firstCleanable;
+
+    private Cleaner() {
+        referenceQueue = new ReferenceQueue<Object>();
+        cleanerThread = new Thread() {
+            @Override
+            public void run() {
+                while(true) {
+                    try {
+                        Reference<? extends Object> ref = referenceQueue.remove();
+                        if(ref instanceof CleanerRef) {
+                            ((CleanerRef) ref).clean();
+                        }
+                    } catch (InterruptedException ex) {
+                        Logger.getLogger(Cleaner.class.getName()).log(Level.SEVERE, null, ex);
+                        break;
+                    } catch (Exception ex) {
+                        Logger.getLogger(Cleaner.class.getName()).log(Level.SEVERE, null, ex);
+                    }
+                }
+            }
+        };
+        cleanerThread.setName("JNA Cleaner");
+        cleanerThread.setDaemon(true);
+        cleanerThread.start();
+    }
+
+    public synchronized CleanerRef register(Object obj, Runnable cleanupTask) {
+        // The important side effect is the PhantomReference, that is yielded
+        // after the referent is GCed
+        return add(new CleanerRef(this, obj, referenceQueue, cleanupTask));
+    }
+
+    private synchronized CleanerRef add(CleanerRef ref) {
+        if(firstCleanable == null) {
+            firstCleanable = ref;
+        } else {
+            ref.setNext(firstCleanable);
+            firstCleanable.setPrevious(ref);
+            firstCleanable = ref;
+        }
+        return ref;
+    }
+
+    private synchronized boolean remove(CleanerRef ref) {
```

Comment:
In theory this could be moved inside the `CleanerRef` class, but I like it out here with the `add()` and `register()` methods.